File: harness/framing.py
"""Binary framing for the CodexOS serial byte stream."""

from __future__ import annotations

import struct
import time
from dataclasses import dataclass

from .serial import SerialConnection, SerialError

MAGIC = b"CXOS"
PROTOCOL_VERSION = 1
MAX_PAYLOAD_SIZE = 16 * 1024 * 1024

_HEADER = struct.Struct("<4sHHII")


class FramingError(RuntimeError):
    """A malformed or incomplete CodexOS frame."""


@dataclass(frozen=True, slots=True)
class Frame:
    message_type: int
    request_id: int
    payload: bytes


def encode_frame(frame: Frame) -> bytes:
    """Encode one frame into its exact wire representation."""
    if not 0 <= frame.message_type <= 0xFFFF:
        raise ValueError("message type must be an unsigned 16-bit value")
    if not 0 <= frame.request_id <= 0xFFFFFFFF:
        raise ValueError("request ID must be an unsigned 32-bit value")
    if len(frame.payload) > MAX_PAYLOAD_SIZE:
        raise FramingError("payload exceeds the 16 MiB version 1 limit")

    header = _HEADER.pack(
        MAGIC,
        PROTOCOL_VERSION,
        frame.message_type,
        frame.request_id,
        len(frame.payload),
    )
    return header + frame.payload


def read_frame(
    connection: SerialConnection,
    timeout_seconds: float = 5.0,
) -> Frame:
    """Read and validate one complete frame from a serial connection."""
    deadline = time.monotonic() + timeout_seconds
    header = _read_exact(connection, _HEADER.size, deadline, "header")
    magic, version, message_type, request_id, payload_length = _HEADER.unpack(header)

    if magic != MAGIC:
        raise FramingError(f"invalid frame magic {magic!r}")
    if version != PROTOCOL_VERSION:
        raise FramingError(f"unsupported protocol version {version}")
    if payload_length > MAX_PAYLOAD_SIZE:
        raise FramingError(
            f"payload length {payload_length} exceeds the 16 MiB version 1 limit"
        )

    payload = _read_exact(connection, payload_length, deadline, "payload")
    return Frame(message_type=message_type, request_id=request_id, payload=payload)


def extract_frame(buffer: bytearray) -> Frame | None:
    """Remove and return one complete frame from an incremental byte buffer."""
    prefix_length = min(len(buffer), len(MAGIC))
    if buffer[:prefix_length] != MAGIC[:prefix_length]:
        raise FramingError(f"invalid frame magic {bytes(buffer[:4])!r}")
    if len(buffer) < _HEADER.size:
        return None

    magic, version, message_type, request_id, payload_length = _HEADER.unpack_from(
        buffer
    )
    if magic != MAGIC:
        raise FramingError(f"invalid frame magic {magic!r}")
    if version != PROTOCOL_VERSION:
        raise FramingError(f"unsupported protocol version {version}")
    if payload_length > MAX_PAYLOAD_SIZE:
        raise FramingError(
            f"payload length {payload_length} exceeds the 16 MiB version 1 limit"
        )

    frame_length = _HEADER.size + payload_length
    if len(buffer) < frame_length:
        return None
    payload = bytes(buffer[_HEADER.size:frame_length])
    del buffer[:frame_length]
    return Frame(message_type=message_type, request_id=request_id, payload=payload)


def _read_exact(
    connection: SerialConnection,
    size: int,
    deadline: float,
    part: str,
) -> bytes:
    data = bytearray()
    while len(data) < size:
        remaining = deadline - time.monotonic()
        if remaining <= 0:
            raise TimeoutError(f"timed out reading frame {part}")
        try:
            data.extend(connection.read(size - len(data), remaining))
        except SerialError as error:
            raise FramingError(f"connection closed while reading frame {part}") from error
    return bytes(data)

File: internal/guest/framing.go
package guest

import (
	"encoding/binary"
	"fmt"
	"io"
)

const (
	ProtocolVersion = uint16(1)
	MaxPayloadSize  = 16 * 1024 * 1024
	frameHeaderSize = 16
)

var frameMagic = [4]byte{'C', 'X', 'O', 'S'}

// Frame is one version 1 message on the guest serial protocol.
type Frame struct {
	MessageType uint16
	RequestID   uint32
	Payload     []byte
}

// FramingError reports malformed or incomplete frame data.
type FramingError struct {
	Reason string
}

func (e *FramingError) Error() string { return e.Reason }

// EncodeFrame returns the exact version 1 wire representation of frame.
func EncodeFrame(frame Frame) ([]byte, error) {
	if len(frame.Payload) > MaxPayloadSize {
		return nil, &FramingError{Reason: "payload exceeds the 16 MiB version 1 limit"}
	}

	encoded := make([]byte, frameHeaderSize+len(frame.Payload))
	copy(encoded[:4], frameMagic[:])
	binary.LittleEndian.PutUint16(encoded[4:6], ProtocolVersion)
	binary.LittleEndian.PutUint16(encoded[6:8], frame.MessageType)
	binary.LittleEndian.PutUint32(encoded[8:12], frame.RequestID)
	binary.LittleEndian.PutUint32(encoded[12:16], uint32(len(frame.Payload)))
	copy(encoded[frameHeaderSize:], frame.Payload)
	return encoded, nil
}

// ReadFrame reads and validates one frame. Transport deadlines and cancellation
// are owned by the caller because io.Reader has no cancellation contract.
func ReadFrame(reader io.Reader) (Frame, error) {
	header := make([]byte, frameHeaderSize)
	if _, err := io.ReadFull(reader, header); err != nil {
		return Frame{}, &FramingError{Reason: fmt.Sprintf("connection closed while reading frame header: %v", err)}
	}
	messageType, requestID, payloadLength, err := decodeHeader(header)
	if err != nil {
		return Frame{}, err
	}

	payload := make([]byte, payloadLength)
	if _, err := io.ReadFull(reader, payload); err != nil {
		return Frame{}, &FramingError{Reason: fmt.Sprintf("connection closed while reading frame payload: %v", err)}
	}
	return Frame{MessageType: messageType, RequestID: requestID, Payload: payload}, nil
}

// ExtractFrame removes one complete frame from the front of an incremental
// buffer. It returns complete=false when the bytes seen so far are a valid
// frame prefix.
func ExtractFrame(buffer *[]byte) (frame Frame, complete bool, err error) {
	data := *buffer
	prefixLength := min(len(data), len(frameMagic))
	for index := range prefixLength {
		if data[index] != frameMagic[index] {
			return Frame{}, false, &FramingError{Reason: fmt.Sprintf("invalid frame magic %q", data[:prefixLength])}
		}
	}
	if len(data) < frameHeaderSize {
		return Frame{}, false, nil
	}

	messageType, requestID, payloadLength, err := decodeHeader(data[:frameHeaderSize])
	if err != nil {
		return Frame{}, false, err
	}
	frameLength := frameHeaderSize + int(payloadLength)
	if len(data) < frameLength {
		return Frame{}, false, nil
	}
	payload := append([]byte(nil), data[frameHeaderSize:frameLength]...)
	*buffer = data[frameLength:]
	return Frame{MessageType: messageType, RequestID: requestID, Payload: payload}, true, nil
}

func decodeHeader(header []byte) (uint16, uint32, uint32, error) {
	if string(header[:4]) != string(frameMagic[:]) {
		return 0, 0, 0, &FramingError{Reason: fmt.Sprintf("invalid frame magic %q", header[:4])}
	}
	version := binary.LittleEndian.Uint16(header[4:6])
	if version != ProtocolVersion {
		return 0, 0, 0, &FramingError{Reason: fmt.Sprintf("unsupported protocol version %d", version)}
	}
	payloadLength := binary.LittleEndian.Uint32(header[12:16])
	if payloadLength > MaxPayloadSize {
		return 0, 0, 0, &FramingError{Reason: fmt.Sprintf("payload length %d exceeds the 16 MiB version 1 limit", payloadLength)}
	}
	return binary.LittleEndian.Uint16(header[6:8]), binary.LittleEndian.Uint32(header[8:12]), payloadLength, nil
}
